// appium-youiengine-driver/build/lib/basedevice.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

class BaseDevice {
  constructor() {}

  closeApp() {
    let msg = `closeApp is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  endSession() {
    let msg = `endSession is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  installApp(appPath) {
    let msg = `installApp is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  isAppInstalled(channelId) {
    let msg = `isAppInstalled is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  launchApp() {
    let msg = `launchApp is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  removeApp(channelId) {
    let msg = `removeApp is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

  startSession(caps) {
    let msg = `startSession is not supported on this platform`;

    _logger.default.errorAndThrow(msg);
  }

}

var _default = BaseDevice;
exports.default = _default;require('source-map-support').install();
