// appium-youiengine-driver/build/lib/yimac.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _basedevice = _interopRequireDefault(require("./basedevice"));

class YiMac extends _basedevice.default {
  constructor() {
    super();
    let caps;
    let shell;
  }

  closeApp() {
    _logger.default.info(`YiMac: Close App`);

    let process_name = this.caps.app.substring(this.caps.app.lastIndexOf("/") + 1);
    this.shell.exec(`killall ${process_name}`);
  }

  endSession() {
    _logger.default.info(`YiMac: End Session`);

    this.closeApp();
  }

  launchApp() {
    _logger.default.info(`YiMac: Launch app`);

    let spawn = require('child_process').spawn,
        ls = spawn(this.caps.app);

    let showXcodeLog = this.caps.showXcodeLog;
    ls.stdout.on('data', function (data) {
      if (showXcodeLog === true && data != null) {
        _logger.default.debug(`Xcode Log Output: ${data.toString()}`);
      }
    });
    ls.stderr.on('data', function (data) {
      if (showXcodeLog === true && data != null) {
        _logger.default.debug(`Xcode Log Error: ${data.toString()}`);
      }
    });
    ls.on('exit', function (code) {
      if (showXcodeLog === true && code != null) {
        _logger.default.debug(`Application exited with code ${code.toString()}`);
      }
    });
  }

  startSession(caps) {
    _logger.default.info(`BlueSky: Start Session`);

    this.caps = caps;
    this.shell = require('shelljs');
    this.closeApp();
    this.launchApp();
  }

}

var _default = YiMac;
exports.default = _default;require('source-map-support').install();
