// appium-remote-debugger/build/lib/remote-debugger-rpc-client.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _logger = _interopRequireDefault(require("./logger"));

var _lodash = _interopRequireDefault(require("lodash"));

var _bplistCreator = _interopRequireDefault(require("bplist-creator"));

var _bplistParser = _interopRequireDefault(require("bplist-parser"));

var _bufferpack = _interopRequireDefault(require("bufferpack"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _remoteDebugger = require("./remote-debugger");

var _uuidJs = _interopRequireDefault(require("uuid-js"));

var _net = _interopRequireDefault(require("net"));

var _remoteDebuggerMessageHandler = _interopRequireDefault(require("./remote-debugger-message-handler"));

var _remoteMessages = _interopRequireDefault(require("./remote-messages"));

class RemoteDebuggerRpcClient {
  constructor(opts = {}) {
    const _opts$host = opts.host,
          host = _opts$host === void 0 ? '::1' : _opts$host,
          _opts$port = opts.port,
          port = _opts$port === void 0 ? _remoteDebugger.REMOTE_DEBUGGER_PORT : _opts$port,
          socketPath = opts.socketPath,
          _opts$specialMessageH = opts.specialMessageHandlers,
          specialMessageHandlers = _opts$specialMessageH === void 0 ? {} : _opts$specialMessageH,
          messageProxy = opts.messageProxy;
    this.host = host;
    this.port = port;
    this.socketPath = socketPath;
    this.messageProxy = messageProxy;
    this.socket = null;
    this.connected = false;
    this.connId = _uuidJs.default.create().toString();
    this.senderId = _uuidJs.default.create().toString();
    this.curMsgId = 0;
    this.received = Buffer.alloc(0);
    this.readPos = 0;
    this.specialMessageHandlers = specialMessageHandlers;
    this.messageHandler = null;
  }

  connect() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      _this.messageHandler = new _remoteDebuggerMessageHandler.default(_this.specialMessageHandlers);

      if (_this.socketPath) {
        if (_this.messageProxy) {
          _logger.default.debug(`Connecting to remote debugger via proxy through unix domain socket: '${_this.messageProxy}'`);

          _this.socket = _net.default.connect(_this.messageProxy);

          _this.socket.once('connect', () => {
            _logger.default.debug(`Forwarding the actual web inspector socket to the proxy: '${_this.socketPath}'`);

            _this.socket.write(JSON.stringify({
              socketPath: _this.socketPath
            }));
          });
        } else {
          _logger.default.debug(`Connecting to remote debugger through unix domain socket: '${_this.socketPath}'`);

          _this.socket = _net.default.connect(_this.socketPath);
        }
      } else {
        if (_this.messageProxy) {
          _this.port = _this.messageProxy;
        }

        _logger.default.debug(`Connecting to remote debugger ${_this.messageProxy ? 'via proxy ' : ''}through TCP: ${_this.host}:${_this.port}`);

        _this.socket = new _net.default.Socket({
          type: 'tcp6'
        });

        _this.socket.connect(_this.port, _this.host);
      }

      _this.socket.setNoDelay(true);

      _this.socket.on('close', () => {
        if (_this.connected) {
          _logger.default.debug('Debugger socket disconnected');
        }

        _this.connected = false;
        _this.socket = null;
      });

      _this.socket.on('end', () => {
        _this.connected = false;
      });

      _this.socket.on('data', _this.receive.bind(_this));

      return yield new _bluebird.default((resolve, reject) => {
        _this.socket.on('connect', () => {
          _logger.default.debug(`Debugger socket connected`);

          _this.connected = true;
          resolve();
        });

        _this.socket.on('error', err => {
          if (_this.connected) {
            _logger.default.error(`Socket error: ${err.message}`);

            _this.connected = false;
          }

          reject(err);
        });
      });
    })();
  }

  disconnect() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (_this2.isConnected()) {
        _logger.default.debug('Disconnecting from remote debugger');

        _this2.socket.destroy();
      }

      _this2.connected = false;
    })();
  }

  isConnected() {
    return this.connected;
  }

  setSpecialMessageHandler(key, errorHandler, handler) {
    this.messageHandler.setSpecialMessageHandler(key, errorHandler, handler);
  }

  getSpecialMessageHandler(key) {
    return this.messageHandler.getSpecialMessageHandler(key);
  }

  setDataMessageHandler(key, errorHandler, handler) {
    this.messageHandler.setDataMessageHandler(key, errorHandler, handler);
  }

  allowNavigationWithoutReload(allow = true) {
    this.messageHandler.allowNavigationWithoutReload(allow);
  }

  selectApp(appIdKey, applicationConnectedHandler) {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      return yield new _bluebird.default((resolve, reject) => {
        let onAppChange = dict => {
          let oldAppIdKey = dict.WIRHostApplicationIdentifierKey;
          let correctAppIdKey = dict.WIRApplicationIdentifierKey;

          if (oldAppIdKey && correctAppIdKey !== oldAppIdKey) {
            _logger.default.debug(`We were notified we might have connected to the wrong app. ` + `Using id ${correctAppIdKey} instead of ${oldAppIdKey}`);
          }

          applicationConnectedHandler(dict);
          reject(new Error('New application has connected'));
        };

        _this3.setSpecialMessageHandler('_rpc_applicationConnected:', reject, onAppChange);

        return (0, _asyncToGenerator2.default)(function* () {
          let _ref2 = yield _this3.send('connectToApp', {
            appIdKey
          }),
              _ref3 = (0, _slicedToArray2.default)(_ref2, 2),
              connectedAppIdKey = _ref3[0],
              pageDict = _ref3[1];

          if (_lodash.default.isEmpty(pageDict)) {
            let msg = 'Empty page dictionary received';

            _logger.default.debug(msg);

            reject(new Error(msg));
          } else {
            resolve([connectedAppIdKey, pageDict]);
          }
        })();
      }).finally(() => {
        _this3.setSpecialMessageHandler('_rpc_applicationConnected:', null, applicationConnectedHandler);
      });
    })();
  }

  send(command, opts = {}) {
    var _this4 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let onSocketError;
      return new _bluebird.default((resolve, reject) => {
        opts = _lodash.default.defaults({
          connId: _this4.connId,
          senderId: _this4.senderId
        }, opts);
        let data = (0, _remoteMessages.default)(command, opts);
        let socketCb = _lodash.default.noop;

        onSocketError = exception => {
          if (_this4.connected) {
            _logger.default.error(`Socket error: ${exception.message}`);
          }

          reject(exception);
        };

        _this4.socket.on('error', onSocketError);

        if (_this4.messageHandler.hasSpecialMessageHandler(data.__selector)) {
          let specialMessageHandler = _this4.getSpecialMessageHandler(data.__selector);

          _this4.setSpecialMessageHandler(data.__selector, reject, function (...args) {
            _logger.default.debug(`Received response from socket send: '${_lodash.default.truncate(JSON.stringify(args), {
              length: 50
            })}'`);

            specialMessageHandler(...args);

            if (this.messageHandler.hasSpecialMessageHandler(data.__selector)) {
              this.setSpecialMessageHandler(data.__selector, null, specialMessageHandler);
            }

            resolve(args);
          }.bind(_this4));
        } else if (data.__argument && data.__argument.WIRSocketDataKey) {
          _this4.curMsgId++;

          const errorHandler = function errorHandler(err) {
            const msg = `Remote debugger error with code '${err.code}': ${err.message}`;
            reject(new Error(msg));
          };

          _this4.setDataMessageHandler(_this4.curMsgId.toString(), errorHandler, value => {
            const msg = _lodash.default.truncate(_lodash.default.isString(value) ? value : JSON.stringify(value), {
              length: 50
            });

            _logger.default.debug(`Received data response from socket send: '${msg}'`);

            _logger.default.debug(`Original command: ${command}`);

            resolve(value);
          });

          data.__argument.WIRSocketDataKey.id = _this4.curMsgId;
          data.__argument.WIRSocketDataKey = Buffer.from(JSON.stringify(data.__argument.WIRSocketDataKey));
        } else {
          socketCb = resolve;
        }

        _logger.default.debug(`Sending '${data.__selector}' message to remote debugger`);

        let plist;

        try {
          plist = (0, _bplistCreator.default)(data);
        } catch (e) {
          let msg = `Could not create binary plist from data: ${e.message}`;

          _logger.default.error(msg);

          return reject(new Error(msg));
        }

        if (_this4.socket && _this4.connected) {
          _this4.socket.cork();

          try {
            _this4.socket.write(_bufferpack.default.pack('L', [plist.length]));

            _this4.socket.write(plist, socketCb);
          } finally {
            _this4.socket.uncork();
          }
        } else {
          let msg = 'Attempted to write data to socket after it was closed!';

          _logger.default.error(msg);

          reject(new Error(msg));
        }
      }).finally(() => {
        _this4.socket.removeListener('error', onSocketError);
      });
    })();
  }

  receive(data) {
    this.received = Buffer.concat([this.received, data]);
    let dataLeftOver = true;

    while (dataLeftOver) {
      let oldReadPos = this.readPos;
      let prefix = this.received.slice(this.readPos, this.readPos + 4);
      let msgLength;

      try {
        msgLength = _bufferpack.default.unpack('L', prefix)[0];
      } catch (e) {
        _logger.default.error(`Buffer could not unpack: ${e}`);

        return;
      }

      this.readPos += 4;

      if (this.received.length < msgLength + this.readPos) {
        this.readPos = oldReadPos;
        break;
      }

      let body = this.received.slice(this.readPos, msgLength + this.readPos);
      let plist;

      try {
        plist = _bplistParser.default.parseBuffer(body);
      } catch (e) {
        _logger.default.error(`Error parsing binary plist: ${e}`);

        return;
      }

      if (plist.length === 1) {
        plist = plist[0];
      }

      var _arr = ['WIRMessageDataKey', 'WIRDestinationKey', 'WIRSocketDataKey'];

      for (var _i = 0; _i < _arr.length; _i++) {
        let key = _arr[_i];

        if (!_lodash.default.isUndefined(plist[key])) {
          plist[key] = plist[key].toString("utf8");
        }
      }

      this.readPos += msgLength;
      let leftOver = this.received.length - this.readPos;

      if (leftOver !== 0) {
        let chunk = Buffer.alloc(leftOver);
        this.received.copy(chunk, 0, this.readPos);
        this.received = chunk;
      } else {
        this.received = Buffer.alloc(0);
        dataLeftOver = false;
      }

      this.readPos = 0;

      if (plist) {
        this.messageHandler.handleMessage(plist);
      }
    }
  }

  setTimelineEventHandler(timelineEventHandler) {
    this.timelineEventHandler = timelineEventHandler;
    this.messageHandler.setTimelineEventHandler(timelineEventHandler);
  }

  setConsoleLogEventHandler(consoleEventHandler) {
    this.consoleEventHandler = consoleEventHandler;
    this.messageHandler.setConsoleLogEventHandler(consoleEventHandler);
  }

  setNetworkLogEventHandler(networkEventHandler) {
    this.networkEventHandler = networkEventHandler;
    this.messageHandler.setNetworkEventHandler(networkEventHandler);
  }

}

exports.default = RemoteDebuggerRpcClient;require('source-map-support').install();
