// appium-adb/build/lib/logcat.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _teen_process = require("teen_process");

var _appiumSupport = require("appium-support");

var _bluebird = _interopRequireDefault(require("bluebird"));

var _events = _interopRequireDefault(require("events"));

const EventEmitter = _events.default.EventEmitter;

const log = _appiumSupport.logger.getLogger('Logcat');

const MAX_BUFFER_SIZE = 10000;
const LOGCAT_PROC_STARTUP_TIMEOUT = 10000;

class Logcat extends EventEmitter {
  constructor(opts = {}) {
    super();
    this.adb = opts.adb;
    this.clearLogs = opts.clearDeviceLogsOnStart || false;
    this.debug = opts.debug;
    this.debugTrace = opts.debugTrace;
    this.maxBufferSize = opts.maxBufferSize || MAX_BUFFER_SIZE;
    this.logs = [];
    this.logIdxSinceLastRequest = 0;
  }

  startCapture() {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      let started = false;
      return yield new _bluebird.default(function () {
        var _ref = (0, _asyncToGenerator2.default)(function* (_resolve, _reject) {
          const resolve = function resolve(...args) {
            started = true;

            _resolve(...args);
          };

          const reject = function reject(...args) {
            started = true;

            _reject(...args);
          };

          if (_this.clearLogs) {
            yield _this.clear();
          }

          log.debug('Starting logcat capture');
          _this.proc = new _teen_process.SubProcess(_this.adb.path, _this.adb.defaultArgs.concat(['logcat', '-v', 'threadtime']));

          _this.proc.on('exit', (code, signal) => {
            log.error(`Logcat terminated with code ${code}, signal ${signal}`);
            _this.proc = null;

            if (!started) {
              log.warn('Logcat not started. Continuing');
              resolve();
            }
          });

          _this.proc.on('lines-stderr', lines => {
            var _iteratorNormalCompletion = true;
            var _didIteratorError = false;
            var _iteratorError = undefined;

            try {
              for (var _iterator = lines[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
                let line = _step.value;

                if (/execvp\(\)/.test(line)) {
                  log.error('Logcat process failed to start');
                  reject(new Error(`Logcat process failed to start. stderr: ${line}`));
                }

                _this.outputHandler(line, 'STDERR: ');
              }
            } catch (err) {
              _didIteratorError = true;
              _iteratorError = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion && _iterator.return != null) {
                  _iterator.return();
                }
              } finally {
                if (_didIteratorError) {
                  throw _iteratorError;
                }
              }
            }

            resolve();
          });

          _this.proc.on('lines-stdout', lines => {
            resolve();
            var _iteratorNormalCompletion2 = true;
            var _didIteratorError2 = false;
            var _iteratorError2 = undefined;

            try {
              for (var _iterator2 = lines[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
                let line = _step2.value;

                _this.outputHandler(line);
              }
            } catch (err) {
              _didIteratorError2 = true;
              _iteratorError2 = err;
            } finally {
              try {
                if (!_iteratorNormalCompletion2 && _iterator2.return != null) {
                  _iterator2.return();
                }
              } finally {
                if (_didIteratorError2) {
                  throw _iteratorError2;
                }
              }
            }
          });

          yield _this.proc.start(0);
          setTimeout(resolve, LOGCAT_PROC_STARTUP_TIMEOUT);
        });

        return function (_x, _x2) {
          return _ref.apply(this, arguments);
        };
      }());
    })();
  }

  outputHandler(output, prefix = '') {
    output = output.trim();

    if (!output) {
      return;
    }

    if (this.logs.length >= this.maxBufferSize) {
      this.logs.shift();

      if (this.logIdxSinceLastRequest > 0) {
        --this.logIdxSinceLastRequest;
      }
    }

    const outputObj = {
      timestamp: Date.now(),
      level: 'ALL',
      message: output
    };
    this.logs.push(outputObj);
    this.emit('output', outputObj);
    const isTrace = /W\/Trace/.test(output);

    if (this.debug && (!isTrace || this.debugTrace)) {
      log.debug(prefix + output);
    }
  }

  stopCapture() {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      log.debug('Stopping logcat capture');

      if (!_this2.proc || !_this2.proc.isRunning) {
        log.debug('Logcat already stopped');
        _this2.proc = null;
        return;
      }

      _this2.proc.removeAllListeners('exit');

      yield _this2.proc.stop();
      _this2.proc = null;
    })();
  }

  getLogs() {
    if (this.logIdxSinceLastRequest < this.logs.length) {
      const result = this.logs.slice(this.logIdxSinceLastRequest);
      this.logIdxSinceLastRequest = this.logs.length;
      return result;
    }

    return [];
  }

  getAllLogs() {
    return this.logs;
  }

  clear() {
    var _this3 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      log.debug('Clearing logcat logs from device');

      try {
        const args = _this3.adb.defaultArgs.concat(['logcat', '-c']);

        log.debug(`Running '${_this3.adb.path} ${args.join(' ')}'`);
        yield (0, _teen_process.exec)(_this3.adb.path, args);
      } catch (err) {
        log.warn(`Failed to clear logcat logs: ${err.message}`);
      }
    })();
  }

}

var _default = Logcat;
exports.default = _default;require('source-map-support').install();
