// appium-android-driver/build/lib/commands/coverage.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _temp = _interopRequireDefault(require("temp"));

var _appiumSupport = require("appium-support");

var _logger = _interopRequireDefault(require("../logger"));

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

async function unlinkFile(file) {
  if (await _appiumSupport.fs.exists(file)) {
    await _appiumSupport.fs.unlink(file);
  }
}

commands.endCoverage = async function endCoverage(intentToBroadcast, ecOnDevicePath) {
  let localFile = _temp.default.path({
    prefix: 'appium',
    suffix: '.ec'
  });

  await unlinkFile(localFile);
  let b64data = '';

  try {
    await this.adb.rimraf(ecOnDevicePath);
    await this.adb.broadcastProcessEnd(intentToBroadcast, this.appProcess);
    await this.adb.pull(ecOnDevicePath, localFile);
    let data = await _appiumSupport.fs.readFile(localFile);
    b64data = Buffer.from(data).toString('base64');
    await unlinkFile(localFile);
  } catch (err) {
    _logger.default.debug(`Error ending test coverage: ${err.message}`);
  }

  return b64data;
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
