// appium-support/build/lib/fs.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.fs = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _fs2 = _interopRequireDefault(require("fs"));

var _rimraf = _interopRequireDefault(require("rimraf"));

var _md5File = _interopRequireDefault(require("md5-file"));

var _ncp = _interopRequireDefault(require("ncp"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _mv = _interopRequireDefault(require("mv"));

var _which = _interopRequireDefault(require("which"));

var _glob = _interopRequireDefault(require("glob"));

var _crypto = _interopRequireDefault(require("crypto"));

const md5 = _bluebird.default.promisify(_md5File.default);

let fs = {
  hasAccess(path) {
    var _this = this;

    return (0, _asyncToGenerator2.default)(function* () {
      try {
        yield _this.access(path, _fs2.default.R_OK);
      } catch (err) {
        return false;
      }

      return true;
    })();
  },

  exists(path) {
    return this.hasAccess(path);
  },

  rimraf: _bluebird.default.promisify(_rimraf.default),

  mkdir(dirName) {
    return (0, _asyncToGenerator2.default)(function* () {
      let _mkdir = _bluebird.default.promisify(_fs2.default.mkdir);

      try {
        yield _mkdir(dirName);
      } catch (err) {
        if (err && err.code !== 'EEXIST') {
          throw err;
        }
      }
    })();
  },

  copyFile(source, destination, ...otherArgs) {
    var _this2 = this;

    return (0, _asyncToGenerator2.default)(function* () {
      if (!(yield _this2.hasAccess(source))) {
        throw new Error(`The file at '${source}' does not exist or is not accessible`);
      }

      return yield _bluebird.default.promisify(_ncp.default)(source, destination, ...otherArgs);
    })();
  },

  md5(filePath) {
    return (0, _asyncToGenerator2.default)(function* () {
      return yield md5(filePath);
    })();
  },

  mv: _bluebird.default.promisify(_mv.default),
  which: _bluebird.default.promisify(_which.default),
  glob: _bluebird.default.promisify(_glob.default),

  hash(filePath, algorithm = 'sha1') {
    return (0, _asyncToGenerator2.default)(function* () {
      return yield new _bluebird.default((resolve, reject) => {
        const fileHash = _crypto.default.createHash(algorithm);

        const readStream = _fs2.default.createReadStream(filePath);

        readStream.on('error', e => reject(new Error(`Cannot calculate ${algorithm} hash for '${filePath}'. Original error: ${e.message}`)));
        readStream.on('data', chunk => fileHash.update(chunk));
        readStream.on('end', () => resolve(fileHash.digest('hex')));
      });
    })();
  }

};
exports.fs = fs;
const simples = ['open', 'close', 'access', 'readFile', 'writeFile', 'write', 'read', 'readlink', 'chmod', 'unlink', 'readdir', 'stat', 'rename', 'lstat'];

for (var _i = 0; _i < simples.length; _i++) {
  const s = simples[_i];
  fs[s] = _bluebird.default.promisify(_fs2.default[s]);
}

const constants = ['F_OK', 'R_OK', 'W_OK', 'X_OK', 'constants'];

for (var _i2 = 0; _i2 < constants.length; _i2++) {
  const c = constants[_i2];
  fs[c] = _fs2.default[c];
}

var _default = fs;
exports.default = _default;require('source-map-support').install();
