// appium-support/build/lib/system.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isWindows = isWindows;
exports.isMac = isMac;
exports.isLinux = isLinux;
exports.isOSWin64 = isOSWin64;
exports.arch = arch;
exports.macOsxVersion = macOsxVersion;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _teen_process = require("teen_process");

var _os = _interopRequireDefault(require("os"));

const VERSION_PATTERN = /^(\d+\.\d+)/m;

function isWindows() {
  return _os.default.type() === 'Windows_NT';
}

function isMac() {
  return _os.default.type() === 'Darwin';
}

function isLinux() {
  return !isWindows() && !isMac();
}

function isOSWin64() {
  return process.arch === 'x64' || process.env.hasOwnProperty('PROCESSOR_ARCHITEW6432');
}

function arch() {
  return _arch.apply(this, arguments);
}

function _arch() {
  _arch = (0, _asyncToGenerator2.default)(function* () {
    if (isLinux() || isMac()) {
      let _ref = yield (0, _teen_process.exec)('uname', ['-m']),
          stdout = _ref.stdout;

      return stdout.trim() === 'i686' ? '32' : '64';
    } else if (isWindows()) {
      let is64 = this.isOSWin64();
      return is64 ? '64' : '32';
    }
  });
  return _arch.apply(this, arguments);
}

function macOsxVersion() {
  return _macOsxVersion.apply(this, arguments);
}

function _macOsxVersion() {
  _macOsxVersion = (0, _asyncToGenerator2.default)(function* () {
    let stdout;

    try {
      stdout = (yield (0, _teen_process.exec)('sw_vers', ['-productVersion'])).stdout.trim();
    } catch (err) {
      throw new Error(`Could not detect Mac OS X Version: ${err}`);
    }

    const versionMatch = VERSION_PATTERN.exec(stdout);

    if (!versionMatch) {
      throw new Error(`Could not detect Mac OS X Version from sw_vers output: '${stdout}'`);
    }

    return versionMatch[1];
  });
  return _macOsxVersion.apply(this, arguments);
}require('source-map-support').install();
