// appium-support/build/lib/tempdir.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.open = open;
exports.path = path;
exports.staticDir = staticDir;
exports.openDir = void 0;

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _fs = _interopRequireDefault(require("./fs"));

var _os = _interopRequireDefault(require("os"));

var _path2 = _interopRequireDefault(require("path"));

var _constants = _interopRequireDefault(require("constants"));

var _logger = _interopRequireDefault(require("./logger"));

const RDWR_EXCL = _constants.default.O_CREAT | _constants.default.O_TRUNC | _constants.default.O_RDWR | _constants.default.O_EXCL;

function tempDir() {
  return _tempDir.apply(this, arguments);
}

function _tempDir() {
  _tempDir = (0, _asyncToGenerator2.default)(function* () {
    let now = new Date();

    let filePath = _path2.default.join(_os.default.tmpdir(), [now.getFullYear(), now.getMonth(), now.getDate(), '-', process.pid, '-', (Math.random() * 0x100000000 + 1).toString(36)].join(''));

    yield _fs.default.mkdir(filePath);
    return filePath;
  });
  return _tempDir.apply(this, arguments);
}

function path(_x, _x2) {
  return _path.apply(this, arguments);
}

function _path() {
  _path = (0, _asyncToGenerator2.default)(function* (rawAffixes, defaultPrefix) {
    let affixes = parseAffixes(rawAffixes, defaultPrefix);
    let name = [affixes.prefix, affixes.suffix].join('');
    let tempDirectory = yield tempDir();
    return _path2.default.join(tempDirectory, name);
  });
  return _path.apply(this, arguments);
}

function open(_x3) {
  return _open.apply(this, arguments);
}

function _open() {
  _open = (0, _asyncToGenerator2.default)(function* (affixes) {
    let filePath = yield path(affixes, 'f-');

    try {
      let fd = yield _fs.default.open(filePath, RDWR_EXCL, 0o600);
      return {
        path: filePath,
        fd
      };
    } catch (err) {
      _logger.default.errorAndThrow(err);
    }
  });
  return _open.apply(this, arguments);
}

function parseAffixes(rawAffixes, defaultPrefix) {
  let affixes = {
    prefix: null,
    suffix: null
  };

  if (rawAffixes) {
    switch (typeof rawAffixes) {
      case 'string':
        affixes.prefix = rawAffixes;
        break;

      case 'object':
        affixes = rawAffixes;
        break;

      default:
        throw new Error(`Unknown affix declaration: ${affixes}`);
    }
  } else {
    affixes.prefix = defaultPrefix;
  }

  return affixes;
}

const _static = tempDir();

const openDir = tempDir;
exports.openDir = openDir;

function staticDir() {
  return _staticDir.apply(this, arguments);
}

function _staticDir() {
  _staticDir = (0, _asyncToGenerator2.default)(function* () {
    return _static;
  });
  return _staticDir.apply(this, arguments);
}require('source-map-support').install();
